### DISCUSSION TITLE
get path to solution filter file (slnf)?

### DISCUSSION THREAD

**Discussion by lsoft** [VsixCommunity/Community.VisualStudio.Toolkit#178] [Q&A]
Hello, in my VSIX I'm trying to determine what solution **filter** is opened now in VS. Looks like there is no correct way to do this.

I'd found 2 ugly workarounds:

# 1 (take the name)

```csharp
var vss = AsyncPackage.GetGlobalService(typeof(SVsSolution)) as IVsSolution;
vss.GetSolutionInfo(out var a, out var b, out var c);
```

where `c` may contains a `slnf` file name in. But it's only a name, not the path. I need it so badly, so I found a second way:

# 2 (take the path)

```csharp
var textBlock = visualStudioMainWindow.GetRecursiveByName("PART_SolutionNameTextBlock"); //get the WPF control
string filePath = (textBlock.TemplatedParent as dynamic).ToolTip.ToString(); //grab its tooltip
```
this workaround provides a full path to `snlf` file,  but looks horrible, and does not work fine in scenarios where the current sln/slnf is changing.

So, any ideas? I'd posted a [ticket](https://developercommunity.visualstudio.com/t/no-way-to-get-path-to-solution-filter-fi/1520237), but get no response at all...

Thanks!

**Comment by reduckted** [comment]
I haven't been able to find anywhere that the solution filter's file name is publicly accessible, but there is a property on the internal `Microsoft.VisualStudio.CommonIDE.Solutions.Solution` type that contains the full path of the solution filter file. That type is what the solution's `IVsHierarchy` is (of course, it's internal, so I guess it could be changed in a future release). If you're OK with using reflection, then you could use this:

```c#
private async Task<string?> GetSolutionFilterFilePath() {
    Solution? solution;

    solution = await VS.Solutions.GetCurrentSolutionAsync();

    if (solution is not null) {
        PropertyInfo? solutionFilterFilePathProperty;

        solution.GetItemInfo(out IVsHierarchy hierarchy, out _, out _);

        solutionFilterFilePathProperty = hierarchy.GetType().GetProperty("SolutionFilterFilePath");

        if (solutionFilterFilePathProperty is not null) {
            return solutionFilterFilePathProperty.GetValue(hierarchy) as string;
        }
    }

    return null;
}
```

**Reply by lsoft** [reply]
@reduckted your workaround is better than mine! Probably the best way is to use both workarounds: if `GetProperty("SolutionFilterFilePath")`-way returns null, then switch to `PART_SolutionNameTextBlock`-way.

Thank you, this information is interesting to know!
